session/session.go

```diff
@@ -346,96 +295,143 @@ func (s *Session) handlePeriodicSearch(ctx context.Context) {
 }
 
 func (s *Session) randomLiveWant() cid.Cid {
-	if len(s.liveWants) == 0 {
+	s.sw.RLock()
+	defer s.sw.RUnlock()
+
+	if len(s.sw.liveWants) == 0 {
 		return cid.Cid{}
 	}
-	i := rand.Intn(len(s.liveWants))
+	i := rand.Intn(len(s.sw.liveWants))
 	// picking a random live want
-	for k := range s.liveWants {
+	for k := range s.sw.liveWants {
 		if i == 0 {
 			return k
 		}
 		i--
 	}
 	return cid.Cid{}
 }
+
 func (s *Session) handleShutdown() {
 	s.idleTick.Stop()
 
-	live := make([]cid.Cid, 0, len(s.liveWants))
-	for c := range s.liveWants {
+	live := s.liveWants()
+	s.wm.CancelWants(s.ctx, live, nil, s.id)
+}
+
+func (s *Session) liveWants() []cid.Cid {
+	s.sw.RLock()
+	defer s.sw.RUnlock()
+
+	live := make([]cid.Cid, 0, len(s.sw.liveWants))
+	for c := range s.sw.liveWants {
 		live = append(live, c)
 	}
-	s.wm.CancelWants(s.ctx, live, nil, s.id)
+	return live
 }
 
-func (s *Session) cidIsWanted(c cid.Cid) bool {
-	_, ok := s.liveWants[c]
+func (s *Session) unlockedIsWanted(c cid.Cid) bool {
+	_, ok := s.sw.liveWants[c]
 	if !ok {
-		ok = s.tofetch.Has(c)
+		ok = s.sw.toFetch.Has(c)
 	}
 	return ok
 }
 
-func (s *Session) processIncoming(ctx context.Context, ks []cid.Cid) {
-	for _, c := range ks {
-		if s.cidIsWanted(c) {
-			// If the block CID was in the live wants queue, remove it
-			tval, ok := s.liveWants[c]
-			if ok {
-				s.latTotal += time.Since(tval)
-				delete(s.liveWants, c)
-			} else {
-				// Otherwise remove it from the tofetch queue, if it was there
-				s.tofetch.Remove(c)
-			}
-			s.fetchcnt++
-
-			// We've received new wanted blocks, so reset the number of ticks
-			// that have occurred since the last new block
-			s.consecutiveTicks = 0
-
-			// Keep track of CIDs we've successfully fetched
-			s.pastWants.Push(c)
-		}
+func (s *Session) handleIncoming(ctx context.Context, rcv rcvFrom) {
+	// Record statistics only if the blocks came from the network
+	// (blocks can also be received from the local node)
+	if rcv.from != "" {
+		s.updateReceiveCounters(ctx, rcv)
 	}
 
-	// Transfer as many CIDs as possible from the tofetch queue into the
-	// live wants queue
-	toAdd := s.wantBudget()
-	if toAdd > s.tofetch.Len() {
-		toAdd = s.tofetch.Len()
-	}
-	if toAdd > 0 {
-		var keys []cid.Cid
-		for i := 0; i < toAdd; i++ {
-			keys = append(keys, s.tofetch.Pop())
-		}
-		s.wantBlocks(ctx, keys)
+	// Update the want list
+	wanted, totalLatency := s.blocksReceived(rcv.ks)
+	if len(wanted) == 0 {
+		return
 	}
+
+	// We've received the blocks so we can cancel any outstanding wants for them
+	s.cancelIncoming(ctx, wanted)
+
+	s.idleTick.Stop()
+
+	// Process the received blocks
+	s.processIncoming(ctx, wanted, totalLatency)
+
+	s.resetIdleTick()
 }
 
 func (s *Session) updateReceiveCounters(ctx context.Context, rcv rcvFrom) {
-	for _, k := range rcv.ks {
-		// Inform the request splitter of unique / duplicate blocks
-		if s.cidIsWanted(k) {
+	s.sw.RLock()
```

Comment:
Are we trying to avoid repeatedly taking the lock? I'm concerned that `Record*Block` blocks. It might be better to just repeatedly retake the lock, check something, then drop it. Or, check everything, store the results in a slice, then record after the fact.

Reply:
On the other hand, do we ever take the _write_ lock anywhere but inside the main loop? In that case, this doesn't really matter.